### PR TITLE
Add Java 17 to build/test matrix

### PR DESCRIPTION
**Pull request by basil**
### Problem

As of #6356 we are shipping core with a `MANIFEST.MF` that contains four (4) `--add-opens` directives that are needed in order to run Jenkins on current releases of XStream with Java 17. Unfortunately, these directives do not take effect during tests, so the tests fail with reflection-related errors.

### Solution

At its core, this PR adds the same `--add-opens` directives to the Surefire JVM arguments when running tests on Java 9+ to enable Java 17 testing. This allows us to add Java 17 to the build/test matrix in the `Jenkinsfile` and improves the project's platform support matrix, establishing a new quality baseline that will make it harder to introduce new incompatibilities by accident in the future.

### Implementation

There are two minor issues with Animal Sniffer and Spotless that I am working around in `pom.xml` pending the removal of Java 8 support; please read the comments and the referenced Jira tickets for more information.

### Testing done

I have been testing this PR successfully on Java 17 over many test runs throughout the past week. As of this PR, the vast majority of core and core components will be built and tested on Java 8, Java 11, and Java 17. 

### Future work

The plugin toolchain still has rough to nonexistent support for anything other than Java 8. I am working on a more comprehensive update to the plugin toolchain that covers Java 11 and Java 17 and will share more information once I have a clear design in mind.

### Proposed changelog entries

Run core test suite on Java 17.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — APPROVED** [review]

File: Jenkinsfile
Comment (jglick) [inline Jenkinsfile:34]: jenkins-infra/helpdesk#2822

File: pom.xml
Comment (jglick) [inline pom.xml:833]: (Well, we _could_ remove it, we choose not to.)

File: pom.xml
Comment (jglick) [inline pom.xml:834]:
```suggestion
          running on Java 17. When https://github.com/jenkinsci/pom/pull/209 is incorporated, this can be deleted.
```
for https://github.com/jenkinsci/pom/pull/209?

File: pom.xml
Comment (jglick) [inline pom.xml:841]: https://issues.jenkins.io/browse/JENKINS-68015 :eyes: workaround not currently applied in https://github.com/jenkinsci/pom/pull/209?

File: test/pom.xml
Comment (jglick) [inline test/pom.xml:232]: Is there a way to define a property in the `jdk-9-and-above` profile that defaults to empty which could be inserted here, so we do not need to duplicate `${jacocoSurefireArgs} -Xmx1g`? I guess it not worth the bother given that this is a temporary situation anyway.

**Review by jglick — APPROVED** [review]
Nice!

File: pom.xml
Comment (basil) [reply pom.xml:833]: I think we ought to keep it as long as we still compile releases using Java 8 (which has no `<release>` support). Hence the comment that we can drop Animal Sniffer when we drop Java 8.

File: pom.xml
Comment (basil) [reply pom.xml:834]: Maybe. I am thinking about doing a more robust implementation of that.

File: pom.xml
Comment (basil) [reply pom.xml:841]: The workaround referenced in [JENKINS-68015 (comment)](https://issues.jenkins.io/browse/JENKINS-68015?focusedCommentId=422939&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-422939) would need to be applied to _this_ repository's `.mvn/jvm.config`, not the parent POM. But doing that in this repository is not feasible as long as we continue to support building with Java 8.

File: test/pom.xml
Comment (basil) [reply test/pom.xml:232]: I do not think that reducing duplication is always a worthy goal, especially if it comes at the cost of additional complexity.

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
